### PR TITLE
basic_auth: support authorization header override

### PR DESCRIPTION
**Pull request by milton0825**
Commit Message: basic_auth: support authorization header override

Additional Description: provide a way to do basic authorization on headers other than `:Authorization`. Say we are doing two level of basic authorization. One at the proxy level with header name `Proxy-Authorization` and the other one at the application level with header name `Authorization`.

Risk Level: low
Testing: unit test
Docs Changes: changelog
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:49]: `authentication_header`?

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.h:62]: Use Http::Lowercase string here.

**Review by wbpcode — COMMENTED** [review]
Thanks for the contribution. Lgtm overall except some minor comment.

@zhaohuabing for a code owner review. :)

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h
Comment (milton0825) [reply source/extensions/filters/http/basic_auth/basic_auth_filter.h]: fixed

**Comment by milton0825** [comment]
@wbpcode this is ready for another look

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:47]:
```suggestion
  // If it is not specified, the filter loads credential from ``Authorization`` header
```

File: changelogs/1.31.0.yaml
Comment (wbpcode) [inline changelogs/1.31.0.yaml:540]: Please add the new api field link here. You can get lots of example in the change log.

**Review by wbpcode — COMMENTED** [review]
Thanks for the update. LGTM overall with only two minor comments. :)

**Comment by milton0825** [comment]
@wbpcode this is ready for another review.

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (zhaohuabing) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:45]:
```suggestion
  // This field specifies the request header to load the basic credential from.
```

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (zhaohuabing) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:47]:
```suggestion
  // If it is not specified, the filter loads the credential from the "Authorization" header.
```

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (zhaohuabing) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:48]: How about naming it `credential_header`?  It seems more precise than `authentication_header` .

**Review by zhaohuabing — COMMENTED** [review]
Thanks for your contribution! I have a few minor suggestions regarding the comments and naming.

**Comment by zhaohuabing** [comment]
Can we also add support for overwriting the `authorization` header in the per-route configuration? This would enable Envoy Gateway to support this feature in the future.

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (milton0825) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:48]: I actually like `authentication_header` since it indicates it is carrying data needed for authentication.

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (milton0825) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto]: fixed

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (milton0825) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto]: fixed

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (milton0825) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto]: fixed

File: changelogs/1.31.0.yaml
Comment (milton0825) [reply changelogs/1.31.0.yaml]: fixed

**Comment by milton0825** [comment]
@zhaohuabing should i do that in a separate PR?

**Comment by milton0825** [comment]
@wbpcode any idea why https://dev.azure.com/cncf/envoy/_build/results?buildId=180089&view=logs&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=f871f0ce-3dc6-5bd2-7916-5ec61dbe723c is failing?
```
/tmp/tmpkc4csx6i/generated/rst/version_history/v1.31/v1.31.0.rst:237: WARNING: undefined label: 'v1.31:envoy_v3_api_field_extensions.filters.http.basic_auth.v3.basicauth.authentication_header'
```.

```
:ref:`forward_username_header <envoy_v3_api_field_extensions.filters.http.basic_auth.v3.BasicAuth.forward_username_header>`
```
seems to just work in https://github.com/envoyproxy/envoy/blob/main/changelogs/1.30.0.yaml#L278

**Comment by zhaohuabing** [comment]
> @zhaohuabing should i do that in a separate PR?

Sure. Thanks！

**Comment by wbpcode** [comment]
> @wbpcode any idea why https://dev.azure.com/cncf/envoy/_build/results?buildId=180089&view=logs&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=f871f0ce-3dc6-5bd2-7916-5ec61dbe723c is failing?
> 
> ```
> /tmp/tmpkc4csx6i/generated/rst/version_history/v1.31/v1.31.0.rst:237: WARNING: undefined label: 'v1.31:envoy_v3_api_field_extensions.filters.http.basic_auth.v3.basicauth.authentication_header'
> ```
> 
> seems to just work for `forward_username_header` in https://github.com/envoyproxy/envoy/blob/main/changelogs/1.30.0.yaml#L278
> 
> ```
> :ref:`forward_username_header <envoy_v3_api_field_extensions.filters.http.basic_auth.v3.BasicAuth.forward_username_header>`
> ```

cc @phlax could you take a look at this?

File: changelogs/1.31.0.yaml
Comment (phlax) [inline changelogs/1.31.0.yaml:538]: why are you changing a historical changelog? the changelogs are immutable you need to add any current changes to `changelogs/current.yaml`

Comment (milton0825) [reply changelogs/1.31.0.yaml]: oops thought that `v1.31` it not released yet

**Comment by milton0825** [comment]
@wbpcode this is ready for another look

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:48]: I suggested an alterantive name because the original header is `authorization`, and naming this field as `authentication_header` might create some confusion. It's not a strong opinion though.

File: changelogs/1.31.0.yaml
Comment (wbpcode) [reply changelogs/1.31.0.yaml]: hah, I didn't check the change log file name... 🤣

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:48]: Authorization header is used to carry authentication information is usual case in practice. So, I am fine to the authentication_header, but credentical_header is also ok for me. It up to you guys.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks. It would be better to get an approval from @zhaohuabing :)

**Review by zhaohuabing — APPROVED** [review]
The naming comment is non-blocking. It's up to @milton0825

**Comment by milton0825** [comment]
i like authentication_header better

**Comment by pxpnetworks** [comment]
Hi guys, i came across this desire to use auth headers other than Authorization like Proxy-Authorization, however for Proxy-Authorization I think the filter should respond with 407 instead of 401 and proxy-authenticate instead of www-authenticate - am i right?

**Pull request merged**
